File: Lab_03/210042174_T04L03B.py
def find_next_seniors(ranks):
    n = len(ranks)
    next_seniors = [-1] * n
    stack = []

    for i in range(n):
        while stack and ranks[stack[-1]] < ranks[i]:
            next_seniors[stack.pop()] = ranks[i]
        stack.append(i)

    return next_seniors


if __name__ == "__main__":
    ranks = []
    print("Enter ranks (terminate with -1):")
    while True:
        rank = input()
        if rank == "-1":
            break
        ranks.append(int(rank))

    next_seniors = find_next_seniors(ranks)

    print("Next seniors by rank:")
    for i in range(len(ranks)):
        print(
            f"Person with rank {ranks[i]} is senior to Person with rank {next_seniors[i] if next_seniors[i] != -1 else 'None'}")

File: cse_4304_lab/Lab_03/210042174_T04L03B.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

vector<int> findNextSeniors(const vector<int> &ranks)
{
    int n = ranks.size();
    vector<int> nextSeniors(n, -1);
    stack<int> st;

    for (int i = 0; i < n; i++)
    {
        while (!st.empty() && ranks[i] > ranks[st.top()])
        {
            nextSeniors[st.top()] = ranks[i];
            st.pop();
        }
        st.push(i);
    }

    while (!st.empty())
    {
        int i = st.top();
        st.pop();
        if (st.empty())
            break;
        int j = st.top();
        nextSeniors[j] = ranks[i];
    }

    return nextSeniors;
}

int main()
{
    while (true)
    {
        vector<int> ranks;
        int rank;
        while (cin >> rank)
        {
            if (rank == -1)
                break;
            ranks.push_back(rank);
        }

        if (ranks.empty())
            break;

        vector<int> nextSeniors = findNextSeniors(ranks);

        for (int i = 0; i < nextSeniors.size(); i++)
        {
            cout << nextSeniors[i];
            if (i < nextSeniors.size() - 1)
                cout << " ";
        }
        cout << endl;
    }

    return 0;
}
